Don't start daemon in userns mode if graphdir inaccessible

estesp opened the pull request:
Fixes #25847 

Warn the user and fail daemon start if the graphdir path has any
elements which will deny access to the remapped root uid/gid.

Docker-DCO-1.1-Signed-off-by: Phil Estes estesp@linux.vnet.ibm.com

estesp committed:
Don't start daemon in userns mode if graphdir inaccessible

Warn the user and fail daemon start if the graphdir path has any
elements which will deny access to the remapped root uid/gid.

Docker-DCO-1.1-Signed-off-by: Phil Estes <estesp@linux.vnet.ibm.com>

estesp commented:
ping @justincormack per your comment in the issue noted--this errors out after checking world "x" (or ownership/group access).

justincormack commented:
LGTM

vdemeester commented:
LGTM 🐮

justincormack commented on pkg/idtools/idtools_unix.go:
```diff
@@ -58,3 +58,29 @@ func mkdirAs(path string, mode os.FileMode, ownerUID, ownerGID int, mkAll, chown
+// CanAccess takes a valid (existing) directory and a uid, gid pair and determines
+// if that uid, gid pair has access (execute bit) to the directory
+func CanAccess(path string, uid, gid int) bool {
+	statInfo, err := system.Stat(path)
```
os.Stat not system.Stat

justincormack commented:
Janky is complaining

estesp commented:
Interesting--looks like I'm the first user of docker/pkg/system.Stat from common code that is built for darwin; will fix

estesp committed:
Add system.Stat support for darwin/macOS

darwin had unbuildable support for our system.Stat() implementation.

Docker-DCO-1.1-Signed-off-by: Phil Estes <estesp@linux.vnet.ibm.com>

vdemeester commented:
LGTM :cat:
